src/inet/mqtt_client.c

```diff
@@ -604,12 +609,40 @@ static int handle_publish(struct mqtt_client_t *self_p,
     }
 
     topic[topic_size] = '\0';
+    qos = ((flags >> 1) & 0x3);
 
-    if (chan_read(self_p->transport.in_p, buf, 2) != 2) {
-        return (-EIO);
-    }
+    log_object_print(self_p->log_object_p,
+                     LOG_DEBUG,
+                     OSTR("QoS: %d, Flags: Flags: 0x%02x.\r\n"),
+                     qos,
+                     flags);
+
+    if (qos == 0) {
+        payload_size = (size - topic_size - 2);
+    } else {
+        /* Read the packet identifier. */
+        if (chan_read(self_p->transport.in_p, buf, 2) != 2) { 
+            return (-EIO);
+        }
+
+        if (qos == 1) {
+            res = write_fixed_header(self_p, MQTT_PUBACK, 0, 2);
+        } else if (qos == 2) { 
+            res = write_fixed_header(self_p, MQTT_PUBREC, 0, 2);
+        } else {
+            res = (-EPROTO);
+        }
+
+        if (res != 0) {
+            return (res);
+        }
 
-    payload_size = (size - topic_size - 4);
+        if (chan_write(self_p->transport.out_p, &buf, 2) != 2) {
```

Comment:
Write ``&buf[0]`` as that is how it is usually written in the code base. There are places in this file with other style, but let's do it right.